src/collector/mod.rs

```diff
@@ -1,253 +1,314 @@
-    fn collect(&mut self, doc: DocId, score: Score) {
-        C::collect(self, doc, score)
+    /// Combines the fruit associated to the collection of each segments
+    /// into one fruit.
+    fn merge_fruits(&self, children: Vec<Self::Fruit>) -> Self::Fruit;
+
+    /// You should not use this method.
+    ///
+    /// Instead, please use [the `Searcher`'s search(...) method](../struct.Searcher.html#method.search).
+    #[doc(hidden)]
+    fn search(&self, searcher: &Searcher, query: &Query) -> Result<Self::Fruit> {
+        let scoring_enabled = self.requires_scoring();
+        let weight = query.weight(searcher, scoring_enabled)?;
+        let mut fruits = Vec::new();
+        for (segment_ord, segment_reader) in searcher.segment_readers().iter().enumerate() {
+            let mut child: Self::Child = self.for_segment(segment_ord as SegmentLocalId, segment_reader)?;
+            let mut scorer = weight.scorer(segment_reader)?;
+            let delete_bitset_opt = segment_reader.delete_bitset();
+            if let Some(delete_bitset) = delete_bitset_opt {
+                scorer.for_each(&mut |doc, score|
+                    if !delete_bitset.is_deleted(doc) {
+                        child.collect(doc, score);
+                    });
+                fruits.push(child.harvest());
+            } else {
+                scorer.for_each(&mut |doc, score| child.collect(doc, score));
+                fruits.push(child.harvest());
+            }
+
+        }
+        Ok(self.merge_fruits(fruits))
     }
 
-    fn requires_scoring(&self) -> bool {
-        C::requires_scoring(self)
+    /// You should not use this method.
+    ///
+    /// Instead, please use [the `Searcher`'s search(...) method](../struct.Searcher.html#method.search).
+    #[doc(hidden)]
+    fn search_multithreads(&self, searcher: &Searcher, query: &Query, num_threads: usize) -> Result<Self::Fruit> {
+        let segment_readers = searcher.segment_readers();
+        let actual_num_threads = (segment_readers.len() + 1).max(num_threads);
+        let scoring_enabled = self.requires_scoring();
+        let weight = query.weight(searcher, scoring_enabled)?;
+        let thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(actual_num_threads)
+            .thread_name(|thread_id|
+                format!("SearchThread-{}", thread_id))
+            .build()
+            .map_err(|_| TantivyError::SystemError("Failed to spawn a search thread".to_string()))?;
+        let segment_fruits: Vec<Self::Fruit> = thread_pool.install(|| {
+            (0..segment_readers.len())
+                .into_par_iter()
+                .map(|segment_ord| {
+                    let segment_reader = &segment_readers[segment_ord];
+                    let mut scorer = weight.scorer(segment_reader).unwrap();
+                    let mut child = self
+                        .for_segment(segment_ord as u32, segment_reader)?;
```

Comment:
Is it worth refactoring this section here and the code in `search` method above into the same method?

Reply:
Did that and removed the confusing implementation in `Collector`